### PR TITLE
IAM 885 OAuth2 client and resource server integration

### PR DESCRIPTION
**Pull request by BarcoMasile**

### CODE REVIEW

**Review by nsklikas — COMMENTED** [review]
Left some comments, I realize that this is a temporary solution and adding cookie support will change the implementation so some comments may not make much sense. 

Feel free to disregard most of my comments, but I think that the handler should  be broken in 2.

File: pkg/authentication/context.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+	tracer trace.Tracer
+	logger logging.LoggerInterface
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	additionalInfo, _ := json.Marshal(map[string]string{"referer": r.Referer(), "remote-address": r.RemoteAddr})
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/context.go:64]: This is because Hydra rejects empty state parameter and needs a certain amount of characters to have enough entropy. I used this two values because it should be more performant than generating pseudorandomic values and because we could actually use those for auditing/logging

File: pkg/authentication/middlewares.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/http/types"
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/tracing"
+)
+
+type Middleware struct {
+	allowListedEndpoints []string
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/middlewares.go:18]: Great suggestion, thanks!

File: pkg/authentication/handlers.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	tracer trace.Tracer
+	logger logging.LoggerInterface
+}
+
+func (a *API) RegisterEndpoints(mux *chi.Mux) {
+	mux.Get("/api/v0/login", a.handleLogin)
+}
+
+func (a *API) handleLogin(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/handlers.go:63]: I like the suggestion, I would prefer doing it in another PR (like when we switch to cookies, since I'll have to change the tests and implementation anyway, and I could spare it now)

File: pkg/authentication/tokens.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+)
+
+type JWKSTokenVerifier struct {
+	clientID string
```
Comment (wood-push-melon) [inline pkg/authentication/tokens.go:18]: Just a **NONBLOCKING** question (probably I miss some knowledge context): why does the verifier have an attribute `clientID`? Did not see where it is used for.

File: pkg/authentication/context.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+func (o *OAuth2Context) Verifier() TokenVerifier {
+	return o.verifier
+}
+
+func NewOAuth2Context(config *Config, getProvider OIDCProviderSupplier, tracer trace.Tracer, logger logging.LoggerInterface) *OAuth2Context {
+	o := new(OAuth2Context)
+	o.tracer = tracer
+	o.logger = logger
+
+	ctx := oidc.InsecureIssuerURLContext(context.Background(), config.issuer)
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/context.go]: Confirmed, you're right. It was actually the leftover. I removed it. Thanks!

File: pkg/authentication/tokens.go (hunk 4 above)
Comment (BarcoMasile) [reply pkg/authentication/tokens.go]: You are right, I put it there for consistency with the other verifier, and with the thought in mind that it could be helpful for next iterations. But I'll remove it, and if needed we'll add it back. Good catch!

Comment (BarcoMasile) [reply pkg/authentication/tokens.go]: I just pushed the change.

File: pkg/authentication/context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nsklikas) [reply pkg/authentication/context.go:63]: Oh this is used for the state, I didn't notice. The state needs to be random, I don't think that this is enough.

I had a look at the library we use and it turns out that it does not validate the value of the state on the callback. We need to add that (and we need to start storing the state in the cookie).

To add to this we should also send a random `nonce` to the provider and the ID Token that we get back must have the same nonce parameter.

**Review by nsklikas — CHANGES_REQUESTED** [review]
Perhaps we should consider using https://github.com/coreos/go-oidc, it is a wrapper around the oauth2 std lib (the one that we use). Jimm went with this one and I think it would be nice if could have a consistent stack.

TBH I don't love either of those implementations

File: pkg/authentication/tokens.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+type claims struct {
+	Audience []string `json:"aud"`
+}
+
+func (u *UserinfoTokenVerifier) VerifyAccessToken(ctx context.Context, rawAccessToken string) (*Principal, error) {
+	_, span := u.tracer.Start(ctx, "authentication.UserinfoTokenVerifier.VerifyAccessToken")
+	defer span.End()
+
+	info, err := u.provider.UserInfo(ctx, oauth2.StaticTokenSource(&oauth2.Token{AccessToken: rawAccessToken}))
```
Comment (shipperizer) [inline pkg/authentication/tokens.go:73]: same here for otelhttp, we will need to add something to the context to make sure it s using a traceable http client

**Review by shipperizer — COMMENTED** [review]
there are quite a lot of layers of abstractions which took me a while to understand but all in all it seems to be doing what it's supposed to

File: pkg/authentication/handlers.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	scopes               []string `validate:"required,dive,required"`
+}
+
+type oauth2Tokens struct {
+	IDToken      string `json:"id_token"`
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+func NewAuthenticationConfig(specs *config.EnvSpec) *Config {
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/handlers.go:36]: I understand the tradeoff, I was trying to simplify the value carrying but I'm open. I'll change it

File: pkg/authentication/interfaces.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+)
+
+type providerVerifier interface {
+	Verify(ctx context.Context, rawIDToken string) (*oidc.IDToken, error)
+}
+
+type TokenVerifier interface {
+	VerifyAccessToken(ctx context.Context, rawAccessToken string) (*Principal, error)
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/interfaces.go:19]: I'm open to drop it since it's clear anyway
I would usually prefer to have names when multiple parameters are present like multiple strings, to allow distinguishing.
Changing it in next revision

File: pkg/authentication/context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/context.go:63]: As much as I agree with the use of the nonce parameter to avoid replay attacks, I consider that a nice improvement, but an improvement nonetheless, so I am glad to work on that, just not in this PR since it's definitely out scope.
About the state, hydra makes it mandatory, it usually isn't. I chose those values to satisfy hydra's requirements.
I would actually use the state parameter to provide a CSRF protection layer, which also would be too much for this first integration.

I am open to change this state value to something pseudo randomic though, and I'm about to do it in next push.
Thanks!

File: pkg/authentication/tokens.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	"fmt"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+)
+
+type JWKSTokenVerifier struct {
```
Comment (shipperizer) [inline pkg/authentication/tokens.go:17]: let's add the monitor here as we might want to add metrics re validation

**Comment by nsklikas** [comment]
> We are using exactly that library (see [go.mod](https://github.com/canonical/identity-platform-admin-ui/blob/270b05951a7d77645fe1c2ea386cd214e0da932c/go.mod)), which provides only the OIDC layer on top of the OAuth2 "layer" provided by a 3rd party library.

You are right, I am blind. I was confused because a lot of the methods/objects come from the std lib

File: pkg/authentication/handlers.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		_ = json.NewEncoder(w).Encode(
+			types.Response{
+				Status:  http.StatusBadRequest,
+				Message: err.Error(),
+			},
+		)
+		return
+	}
+
+	rawIDToken := oauth2Token.Extra("id_token").(string)
```
Comment (nsklikas) [inline pkg/authentication/handlers.go:87]: We should verify the ID token (similar to https://github.com/coreos/go-oidc/blob/v3/example/idtoken/app.go#L103)

Comment (BarcoMasile) [reply pkg/authentication/handlers.go:87]: I'll do it the next push (coming up), thanks

File: pkg/authentication/context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/context.go:63]: I changed my mind and implemented the nonce check in this PR, so I'll push the changes as soon as I'm done addressing all other comments. Thanks!

File: pkg/authentication/middlewares.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		if m.isAllowListed(r) {
+			next.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		rawAccessToken, err := getBearerToken(r.Header)
+		if err != nil {
+			m.unauthorizedResponse(w, err)
+			return
+		}
```
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/middlewares.go:55]: If that's the plan then we should talk mroe about this, since with the cookie based authentication we could rely on multiple pieces of info (id, access and refresh tokens), so we could (for users) refresh the access token if it's expired.
Let's talk about this and provide a plan for the future

File: pkg/authentication/context.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	o := new(OAuth2Context)
+	o.tracer = tracer
+	o.logger = logger
+
+	provider, err := getProvider(context.Background(), config.issuer)
+	if err != nil {
+		o.logger.Fatalf("Unable to fetch provider info, error: %v", err.Error())
+	}
+
+	var verifier TokenVerifier
```
Comment (wood-push-melon) [inline pkg/authentication/context.go:104]: Just a **NONBLOCKING** question: do you think it's better to move this token verifier "factory" logic to `tokens.go`? It seems that `OAuth2Context` does not need to care about how to create an appropriate token verifier. It only needs to use the factory to get the right one.

File: internal/config/specs.go [hunk 13]
```diff
@@ -23,6 +23,14 @@ type EnvSpec struct {
 	HydraAdminURL       string `envconfig:"hydra_admin_url" required:"true"`
 	OathkeeperPublicURL string `envconfig:"oathkeeper_public_url" required:"true"`
 
+	AuthenticationEnabled           bool     `envconfig:"authentication_enabled" default:"false" validate:"required"`
+	OIDCIssuer                      string   `envconfig:"oidc_issuer" validate:"required"`
+	OAuth2ClientId                  string   `envconfig:"oauth2_client_id" validate:"required"`
+	OAuth2ClientSecret              string   `envconfig:"oauth2_client_secret" validate:"required"`
+	OAuth2RedirectURI               string   `envconfig:"oauth2_redirect_uri" validate:"required"`
+	OAuth2CodeGrantScopes           []string `envconfig:"oauth2_codegrant_scopes" default:"openid,offline_access" validate:"required"`
+	AccessTokenVerificationStrategy string   `envconfig:"access_token_verification_strategy" default:"jwks" validate:"oneof=jwks userinfo"`
```
Comment (nsklikas) [inline internal/config/specs.go:32]: Should we have a common prefix for the oauth2 vars?

File: pkg/authentication/context.go (hunk 12 above)
Comment (BarcoMasile) [reply pkg/authentication/context.go:104]: You are right and that was my original implementation, but since the factory process is so simple, I removed the factory classt and just made it part of the OAuth2Context, so it's one less object to crowd the implementation.
We can still extract it in case things get more complicated.

File: internal/config/specs.go (hunk 13 above)
Comment (BarcoMasile) [reply internal/config/specs.go:32]: I tried to keep it consistent, thing is, the only one missing the oauth2 prefix is the verification strategy (which already has a long name) so I left it like that since "access token" is already explicitly oauth2 related.
OIDC issuer I prefer to keep it OIDC prefixed for its meaning, and the first AuthenticationEnabled I didn't put the prefix there because authn is at a higher conceptual level that OAuth2.
So basically I could add the prefix to the verification strategy making it `OAuth2AccessTokenVerificationStrategy`, I'm open to it

File: pkg/authentication/context.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (BarcoMasile) [reply pkg/authentication/context.go:63]: In the PR change coming up today you'll see the nonce implementation. Thing is, nonce should be encrypted, but encryption will come with next story. So for now it will be base64 encoded, and a todo is there to track that (it will be addressed very soon so no need for a new issue).

File: pkg/authentication/cookies.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	NONCE_COOKIE_NAME = "nonce"
+)
+
+func SetNonceCookie(w http.ResponseWriter, nonce string) {
+	expires := time.Now().Add(2 * time.Minute)
+	http.SetCookie(w, &http.Cookie{
+		Name: NONCE_COOKIE_NAME,
+		// TODO @barco: encrypt this value when cookie encryption is added to the codebase
+		Value:    nonce,
+		Path:     "/api/v0/login",
```
Comment (nsklikas) [inline pkg/authentication/cookies.go:22]: I may be wrong, but I am not sure if this will work when running under a path, e.g. the login URL may be:
`https://example.com/iam-admin/api/v0/login`, afaict in that case this cookie path won't work

File: pkg/authentication/context.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+	monitor monitoring.MonitorInterface
+	rand    *rand.Rand
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	nonce, _ := nonce(r)
+	state := o.state(stateSize)
```
Comment (nsklikas) [inline pkg/authentication/context.go:84]: Since cookie support was added for `nonce`, shouldn't we add the same verification logic for `state`?

File: pkg/authentication/context.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+
+	SetNonceCookie(w, nonce)
+
+	http.Redirect(
+		w, r,
+		o.client.AuthCodeURL(state, oidc.Nonce(nonce), oauth2.SetAuthURLParam("audience", o.client.ClientID)),
```
Comment (nsklikas) [inline pkg/authentication/context.go:90]: Can we add a TODO to remove this once https://github.com/canonical/identity-platform-login-ui/issues/244 is fixed?

File: pkg/authentication/context.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+func (o *OAuth2Context) state(size int) string {
+	state := make([]rune, size)
+	for i := range state {
+		state[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(state)
+}
+
+func nonce(r *http.Request) (string, error) {
+	marshal, err := json.Marshal(map[string]string{"referer": r.Referer(), "remote-address": r.RemoteAddr})
```
Comment (nsklikas) [inline pkg/authentication/context.go:104]: Since we have a random generator for state, why don't we reuse that for nonce as well? Unless we have a use for these values.

File: pkg/authentication/cookies.go [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+	"time"
+)
+
+var (
+	EPOCH             = time.Unix(0, 0)
+	NONCE_COOKIE_NAME = "nonce"
+)
+
+func SetNonceCookie(w http.ResponseWriter, nonce string) {
+	expires := time.Now().Add(2 * time.Minute)
```
Comment (nsklikas) [inline pkg/authentication/cookies.go:17]: I feel like this should be larger, maybe we should make it configurable?

File: pkg/authentication/cookies.go (hunk 14 above)
Comment (BarcoMasile) [reply pkg/authentication/cookies.go:22]: You're not wrong. Right now every handler endpoint is hardcoded and does not take into account any prefix. When we implement that, we will have to take care of that in all the codebase.

File: pkg/authentication/tokens.go (hunk 9 above)
Comment (BarcoMasile) [reply pkg/authentication/tokens.go:19]: Done in both token verifiers

File: pkg/authentication/tokens.go (hunk 6 above)
Comment (BarcoMasile) [reply pkg/authentication/tokens.go:78]: Done for every `context` use in the OAuth2 implementation. The httpclient with Otel transport is shared.

**Review by nsklikas — COMMENTED** [review]
I understand that this is not a complete set of changes and some issues with this codebase will be addressed on later PRs. I think it would make approving this easier if we could have a list of things that need to be done once this PR is merged

File: pkg/authentication/cookies.go (hunk 18 above)
Comment (BarcoMasile) [reply pkg/authentication/cookies.go:17]: Making it configurable right now

File: pkg/authentication/cookies.go (hunk 14 above)
Comment (shipperizer) [reply pkg/authentication/cookies.go:22]: ingress controller will take care of doing the path rewriting, now do we actually care in this case what s inside the cookie given that the path of the request will be rewritten to what we know?

File: pkg/authentication/context.go (hunk 15 above)
Comment (BarcoMasile) [reply pkg/authentication/context.go:85]: I would honestly skip this for now, since I already adeded that check it should be more than enough for a first implementation. We can add/change that whenever we want. My idea would be to make it part of a CSRF protection implementation we can add in the future.

File: pkg/authentication/cookies.go (hunk 14 above)
Comment (BarcoMasile) [reply pkg/authentication/cookies.go:22]: We could "ignore" the path although it's nice to keep this cookie "scoped" to the only endpoint that needs it.
My proposal is: leave it as is. With cookie encryption implemented, we would have our secret key as an additional strong layer of security and could decide to relax the `path` constraint if it troubles us with the deployment

File: pkg/authentication/context.go (hunk 17 above)
Comment (BarcoMasile) [reply pkg/authentication/context.go:106]: I would skip this additional change since it would require me to bring a bunch of changes to make it testable, and would cause an avalanche effect from addressing comments

File: pkg/authentication/context.go (hunk 15 above)
Comment (nsklikas) [reply pkg/authentication/context.go:85]: I have no problem not to add this in this PR to unblock development, but we *need* to add state verification eventually. 

Several attacks described in https://datatracker.ietf.org/doc/html/rfc6819 can be mitigated with the `state` parameter.

File: pkg/authentication/context.go (hunk 17 above)
Comment (nsklikas) [reply pkg/authentication/context.go:106]: Same here. I have no problem with keeping this as it is for now, as long as we have a clear backlog of what needs to be done and that we eventually address this.

We should use random values for `nonce` and `state`. Using headers like `Referer` and `RemoteAddr` opens the window for attacks, because it is very easy for someone to tamper with them

File: pkg/authentication/context.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+const stateSize = 18
+
+var (
+	letterRunes         = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	PrincipalContextKey principalContextKey
+	otelHTTPClient      = http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
+)
+
+type Principal struct {
+	Subject string `json:"sub"`
```
Comment (nsklikas) [inline pkg/authentication/context.go:34]: Question out of scope for this PR: Shouldn't we use the user's email instead of the `sub`? This would be needed for the exposing the user's ID in the UI

File: pkg/authentication/context.go (hunk 17 above)
Comment (BarcoMasile) [reply pkg/authentication/context.go:106]: [Issue](https://github.com/canonical/identity-platform-admin-ui/issues/321) for this

File: pkg/authentication/context.go (hunk 15 above)
Comment (BarcoMasile) [reply pkg/authentication/context.go:85]: Totally agree on adding state check in the near future.
[Issue](https://github.com/canonical/identity-platform-admin-ui/issues/322) to track this.

File: pkg/web/router.go [hunk 20]
```diff
@@ -132,6 +137,16 @@ func NewRouter(config *RouterConfig, wpool pool.WorkerPoolInterface) http.Handle
...
 	}).(*chi.Mux)
 
+	var oauth2Context *authentication.OAuth2Context
+
+	if oauth2Config.Enabled {
+		oauth2Context = authentication.NewOAuth2Context(config.oauth2, oidc.NewProvider, tracer, logger, monitor)
+
+		authenticationMiddleware := authentication.NewAuthenticationMiddleware(oauth2Context, tracer, logger)
+		authenticationMiddleware.SetAllowListedEndpoints("/api/v0/login", "/api/v0/status", "api/v0/metrics")
+		apiRouter = apiRouter.With(authenticationMiddleware.OAuth2Authentication).(*chi.Mux)
```
Comment (nsklikas) [inline pkg/web/router.go:147]: I switched the other routers to use `Use`, should we change this one as well? (the issue that we had with `Use` does not happen with subrouters)

**Review by nsklikas — DISMISSED** [review]
minor comment, but approved

**Review by nsklikas — APPROVED** [review]
